README.md
# d3by5-pie-chart
The d3by5 is part of the d3by5 graph tools, this specific package will draw a pie chart based on Mike Bostocks [Towards Reusable Charts](https://bost.ocks.org/mike/chart/) using d3 v4.0

## NOTE
This is an internal project, you are probably better off using somethin like [C3](https://github.com/c3js/c3). That said, just give it a try and contact us back (no capslock please).

## USAGE

### NPM
Build the project by running
```bash
npm run build
```

When building two versions are built in the dist folder.
use the [version] version if you want to keep a specific version, use submodules or symlinks (or download) of the one without version if you want to update the lib without updating your code.

### Requirejs
Just require the sourcefile, it expects the following files accessible at root so you may have to add a shim for it
* d3by5-base-chart
* d3by5-transitions
* d3 - v3x
* underscore


#### dist/
A browserified and uglified version with all dependencies included. Use this if you only want a simple graph to test.
* pie-chart-[version].min.js    // uglified and minified with version
* pie-chart.min.js              // uglified and minified without version

## DEPENDENCIES
Two dependencies in package.json
* Underscore
* d3

```
npm install
```
will get you all you need

## API
### Required
* width  - Number: the height of the chart
* height - Number: the width of the chart
* data   - Array: the data that produces the chart, [{label: String, value: Number}, {xx}]

### Optional
* fillColor - String/hex: the fillcolor (defaults to 'coral')
* padding - Number: the padding to use (detaults to 2)


## EXAMPLE
All methods are chained, you can simply instanciate a new pie chart like this

```javascript
var pie = require('d3by5-pie-chart');

var piechart = piechart()
                    .width(500)
                    .height(400)
                    .data([{label:'coffee', values: 509}, {label:'tea', values: 1}]);

d3.select('.js-pie-chart')
    .call(piechart);
```

## LICENCE
[MIT](https://opensource.org/licenses/MIT)

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres (well, maybe not quite) to [Semantic Versioning](http://semver.org/).

## [unreleased]

### Added
- Added examples with requirejs implementation

### Changed
- Changed the module loading to UMD

### Fixed
- Fixed an issue with duplicate scipt props in packege.json preventing building
- Fixed an issue where few slices offsets labels in the worng direction


## [0.1.0] - 2016-05-30
### Added
- Added README with initial API definition
- Added submodule 'base' with base graph
- Added initial 'pie-chart' code for testing
- Added build script
- Added aliasify and shimming of base-chart
- Added proper creation of a pie chart from data

### Changed
- Updated README
- Simplify building by removing dependencies in build.sh
- Remove dependencies and inline all graph menthods

### Removed
- Removed submodule 'base' with base graph


[//]: ##############################################
<!---
[//]: # (Legend)
[Added]:        <> (for new features.)
[Changed]:      <> (for changes in existing functionality.)
[Deprecated]:   <> (for once-stable features removed in upcoming releases.)
[Removed]:      <> (for deprecated features removed in this release.)
[Fixed]:        <> (for any bug fixes.)
[Security]:     <> (to invite users to upgrade in case of vulnerabilities.)
--->

src/pie-chart.js
/*!
 * Pie charts
 *
 */
(function (root, factory) {
    if (typeof define === 'function' && define.amd) {
        // AMD. Register as an anonymous module.
        define(['underscore', 'd3', './base-chart', './transitions'] , factory);
    } else if (typeof module === 'object' && module.exports) {
        // Node. Does not work with strict CommonJS, but
        // only CommonJS-like environments that support module.exports,
        // like Node.
        module.exports = factory(require('underscore'), require('d3'), require('./base-chart'), require('./transitions'));
    } else {
        // Browser globals (root is window)
        root.returnExports = factory(root._, root.d3, root.d3by5BaseChart, root.d3by5Transitions);
    }
}(this, function (_, d3, base, transitions) {

'use:strict';


/**
 * The entrypoint
 * @return {[type]} [description]
 */
function pieChart () {

  var chart = {

    type: 'pie',

    options: {
        padding: 2,
        size: 100,
        innerRadius: 0,
        transitionDuration: 1000,
        textStyle : {  // probably tmp until implemented themeing
          'font-family' : '"Helvetica Neue", Arial, Helvetica, sans-serif',
          'font-size'   : 13,
          'fill'        : '#444'
        }
    },

    init: function (selection) {

      if (arguments.length) {
        this.selection = selection;
        this.draw(selection);
      }
      return this;
    },

    draw: function () {
      var width = this.options.width
        , height = this.options.height
        , adjustedHeight = (height - (this.options.margin.top + this.options.margin.bottom)) / 2
        , adjustedWidth = (width - (this.options.margin.left + this.options.margin.right)) / 2
        , radius =  Math.min(adjustedHeight, adjustedWidth)
        , innerRadius = this.options.innerRadius

      ;

      if (this.options.size && this.options.size !== 100) {
        radius = radius * (this.options.size / 100);
        innerRadius = innerRadius * (this.options.size / 100);
      }


      this.radius = radius;
      this.innerRadius = innerRadius;

      this.arc = d3.svg.arc()
              .outerRadius(radius)
              .innerRadius(innerRadius); 

      this.pie = d3.layout.pie()
              .sort(null)
              .value(function(d) {
                return d.values;
              });

      // remove old
      if (this.svg) {
        this.svg.remove();
      }

      this.svg = this.selection.append("svg")
                      .attr("width", width)
                      .attr("height", height)
                      .append("g")
                        .attr("transform", "translate(" + (this.options.margin.left + adjustedWidth) + "," + (this.options.margin.top + adjustedHeight) + ")");

      this.render();

    },


    render: function () {
      // If there are any transitions
      // hijack the data update function
      if (this.options.transition && this.options.transition.t) {
        this.options.transition.t();
      }


      var path
        , that = this
        , drawEvent
        , arcTween = function (a) {
            var i = d3.interpolate(this._current, a);
            this._current = i(0);
            return function(t) {
              return that.arc(i(t));
            };
          }; // redraw the arcs

      //data join
      this.arcs = this.svg
                        .selectAll("path")
                        .data(this.pie(this.options.data), function (d) {
                          return d.data.label;
                        });

      // update, transition
      this.arcs.transition()
          .duration(1000)
          .attrTween("d", arcTween);

      // create new elements
      this.arcs.enter()
                .append('g')
                .attr('class', 'js-pie__path')
                .append("path")
                .attr("fill", function(d) {
                  return that.options.fillColor(d.data.label);
                })
                .attr("d", that.arc)
                .transition()
                .duration(1000)
                .attrTween("d", arcTween)
                .each(function(d) {
                  this._current = d;
                }); // store the initial angles

      // remove unused data
      this.arcs.exit()
          .transition()
          .attrTween("d", arcTween)
          .remove();

      if (this.options.labelPosition !== 'none') {
        this.drawLabels();
      }
      if (this.options.valuesPosition !== 'none') {
        this.drawValues();
      }

      // apply any events that was unbound
      this.applyEvents();

      drawEvent = _.find(this.options.on, function (o) {
        return o.action === 'draw';
      });

      if (drawEvent && _.isFunction (drawEvent.method)) {
        drawEvent.method.call(this);
      }
    },


    drawLabels: function () {

      var labelMultiplier
        , that = this
        , slices
        , diff = this.radius - this.innerRadius
      ;

      labelMultiplier = 1.2 + ((diff/this.radius));



      slices = this.arcs
                  .append('text')
                  .style(this.options.textStyle)
                  .attr("text-anchor", function (d) {
                    // if slightly smaller than a half circle
                    if (d.startAngle < 3) {
                      return 'start';
                    }
                    if ((d.startAngle > 3.3 && d.endAngle < 6) || (d.endAngle - d.startAngle > 0.5)) {
                      return 'end';
                    }
                    return 'middle';
                  })
                  .attr('transform', function (d, i) {
                    var c = that.arc.centroid(d, i);
                    return 'translate(' + (c[0] * labelMultiplier) + ',' + (c[1] * labelMultiplier) +')';
                    // var _obj = {endAngle: d.endAngle,
                    //             innerRadius: that.radius + 50,
                    //             outerRadius: that.radius + 245,
                    //             padAngle: d.padAngle,
                    //             startAngle: d.startAngle};
                    // return "translate(" + that.arc.centroid(_obj, i) + ")";
                  })
                  .text(function (d) {
                    // if (_.isFunction (that.options.labelFormat)) {
                    //   return that.options.labelFormat(d.data.label);
                    // }
                    return d.data.label;
                  });
    },

    drawValues: function () {
      var that = this;

      this.arcs.filter(function(d) {
                        return d.endAngle - d.startAngle > 0.2; })
              .append('text')
            .attr('dy', '.35em')
            .attr('text-anchor', 'middle')
            .attr('transform', function(d, i) { //set the label's origin to the center of the arc
              var _obj = {endAngle: d.endAngle,
                          innerRadius: that.radius/2,
                          outerRadius: that.radius,
                          padAngle: d.padAngle,
                          startAngle: d.startAngle};

              return 'translate(' + that.arc.centroid(_obj, i) + ')rotate(' + that.angle(_obj) + ')';
            })
            .style(this.options.textStyle)
            .style('fill', '#fff')  // override default text color
            .style('font-weight', 'bold')  // override default text weight
      .text(function(d) {
        return that.formatNumber(d.data.values, that.options.valuesFormat);
      });
    },

        // Computes the angle of an arc, converting from radians to degrees.
    angle: function(d) {
      var a = (d.startAngle + d.endAngle) * 90 / Math.PI - 90;
      return a > 90 ? a - 180 : a;
    },

    /**
     * Sets the innerradius (a innerradius > 0 creates a donut)
     * @param  {Number} value - the innerRadius of the chart
     * @return {Mixed}        - the value or chart
     */
    innerRadius: function (value) {
      if (!arguments.length) return this.options.innerRadius;
      this.options.innerRadius = value;
      return chart;
    },

    /**
     * Sets the size, this is a percentage modifier, and should normally not exceede 100, but of cource it is possible
     * @param  {Number} value - the percent to add to the calculated size
     * @return {Mixed}        - the value or chart
     */
    size: function (value) {
      if (!arguments.length) return this.options.size;
      this.options.size = value;
      return chart;
    },

    applyEvents: function () {
      var that = this
        , path = this.svg
                .selectAll("path");
      if (path) {
        _.each(this.options.on, function (value, key) {
          path.on(key, value);
        });
      }
    }

  };
  chart.options = _.extend(chart.options, base.options);
  chart = _.extend(chart, _.omit(base, 'options'), transitions);

  return (chart.init());
}
return pieChart;
}));

src/examples/requirejs/app.js
// Place third party dependencies in the lib folder
//
// Configure loading modules from the lib directory,
// except 'app' ones,

//
// We add paths to everything here as it is a test
//
//
requirejs.config({
    paths: {
        'pie-chart': '../../../src/pie-chart',
        'd3': 'lib/d3',
        'underscore': 'lib/underscore-1.8.3',
        'base-chart': 'lib/base-chart',
        'transitions': 'lib/transitions',
        'base-chart-utils': 'lib/base-chart-utils',
        'base-chart-axis': 'lib/base-chart-axis',
    },
});


require(['pie-chart','underscore','d3'], function (piechart, _, d3) {
  'use strict';
var singleData = [{"values": 864,"label": "Afghanistan"},
                  {"values": 2,"label": "Albania"},
                  {"values": 36,"label": "Algeria"},
                  {"values": 2,"label": "Azerbaijan"},
                  {"values": 552,"label": "Bangladesh"},
                  {"values": 1,"label": "Benin"},
                  {"values": 4,"label": "Burkina Faso"},
                  {"values": 1,"label": "Cameroon"},
                  {"values": 1,"label": "Chad"},
                  {"values": 5,"label": "Côte D'ivoire"},
                  {"values": 827,"label": "Egypt"},
                  {"values": 1161,"label": "Eritrea"},
                  {"values": 84,"label": "Ethiopia"},
                  {"values": 242,"label": "Gambia"},
                  {"values": 15,"label": "Ghana"},
                  {"values": 25,"label": "Guinea"},
                  {"values": 6,"label": "Guinea-Bissau"},
                  {"values": 74,"label": "India"},
                  {"values": 85,"label": "Iran"},
                  {"values": 124,"label": "Iraq"},
                  {"values": 1,"label": "Jamaica"},
                  {"values": 1,"label": "Latvia"},
                  {"values": 1,"label": "Liberia"},
                  {"values": 20,"label": "Libya"},
                  {"values": 184,"label": "Mali"},
                  {"values": 79,"label": "Morocco"},
                  {"values": 1,"label": "Nepal"},
                  {"values": 3,"label": "Niger"},
                  {"values": 259,"label": "Nigeria"},
                  {"values": 1198,"label": "Pakistan"},
                  {"values": 32,"label": "Palestine"},
                  {"values": 0,"label": "Romania"},
                  {"values": 1,"label": "Russia"},
                  {"values": 48,"label": "Senegal"},
                  {"values": 4,"label": "Sierra Leone"},
                  {"values": 1292,"label": "Somalia"},
                  {"values": 17,"label": "Sudan"},
                  {"values": 393,"label": "Syria"},
                  {"values": 4,"label": "Togo"},
                  {"values": 2147,"label": "Tunisia"},
                  {"values": 48,"label": "Turkey"},
                  {"values": 7,"label": "Ukraine"},
                  {"values": 1,"label": "Zimbabwe"}]
    , groupedData = [{"label":"09-2006","values":[
                                                  {"label":"Syria","values":308},
                                                  {"label":"Afghanistan","values":324}
                                                  ]},
                    {"label":"10-2006","values":[
                                                  {"label":"Afghanistan","values":460},
                                                  {"label":"Syria","values":396}
                                                  ]},
                    {"label":"11-2006","values":[
                                                  {"label":"Syria","values":344},
                                                  {"label":"Afghanistan","values":328}
                                                  ]},
                    {"label":"12-2006","values":[
                                                  {"label":"Afghanistan","values":588},
                                                  {"label":"Syria","values":344}
                                                  ]},
                    {"label":"01-2007","values":[
                                                  {"label":"Afghanistan","values":412},
                                                  {"label":"Syria","values":436}
                                                  ]},
                    {"label":"02-2007","values":[
                                                  {"label":"Syria","values":252},
                                                  {"label":"Afghanistan","values":372}
                                                  ]},
                    {"label":"03-2007","values":[
                                                  {"label":"Afghanistan","values":296},
                                                  {"label":"Syria","values":288}
                                                  ]},
                    {"label":"04-2007","values":[
                                                  {"label":"Syria","values":252},
                                                  {"label":"Afghanistan","values":284}
                                                  ]},
                    {"label":"05-2007","values":[
                                                  {"label":"Afghanistan","values":280},
                                                  {"label":"Syria","values":284}
                                                  ]},
                    {"label":"06-2007","values":[
                                                  {"label":"Syria","values":212},
                                                  {"label":"Afghanistan","values":328}
                                                  ]},
                    {"label":"07-2007","values":[
                                                  {"label":"Afghanistan","values":296},
                                                  {"label":"Syria","values":232}
                                                  ]},
                    {"label":"08-2007","values":[
                                                  {"label":"Syria","values":424},
                                                  {"label":"Afghanistan","values":272}
                                                  ]},
                    {"label":"09-2007","values":[
                                                  {"label":"Afghanistan","values":240},
                                                  {"label":"Syria","values":416}
                                                  ]},
                    {"label":"10-2007","values":[
                                                  {"label":"Syria","values":492},
                                                  {"label":"Afghanistan","values":308}
                                                  ]},
                    {"label":"11-2007","values":[
                                                  {"label":"Afghanistan","values":332},
                                                  {"label":"Syria","values":504}
                                                  ]},
                    {"label":"12-2007","values":[
                                                  {"label":"Syria","values":384},
                                                  {"label":"Afghanistan","values":332}
                                                  ]},
                    {"label":"01-2008","values":[
                                                  {"label":"Afghanistan","values":488},
                                                  {"label":"Syria","values":428}
                                                  ]},
                    {"label":"02-2008","values":[
                                                  {"label":"Syria","values":472},
                                                  {"label":"Afghanistan","values":372}
                                                  ]},
                    {"label":"03-2008","values":[
                                                  {"label":"Afghanistan","values":184},
                                                  {"label":"Syria","values":368}
                                                  ]},
                    {"label":"04-2008","values":[
                                                  {"label":"Syria","values":428},
                                                  {"label":"Afghanistan","values":260}
                                                  ]},
                    {"label":"05-2008","values":[
                                                  {"label":"Afghanistan","values":244},
                                                  {"label":"Syria","values":352}
                                                  ]},
                    {"label":"06-2008","values":[
                                                  {"label":"Syria","values":372},
                                                  {"label":"Afghanistan","values":308}
                                                  ]},
                    {"label":"07-2008","values":[
                                                  {"label":"Afghanistan","values":332},
                                                  {"label":"Syria","values":516}
                                                  ]},
                    {"label":"08-2008","values":[
                                                  {"label":"Syria","values":484},
                                                  {"label":"Afghanistan","values":620}
                                                  ]},
                    {"label":"09-2008","values":[
                                                  {"label":"Afghanistan","values":836},
                                                  {"label":"Syria","values":476}
                                                  ]},
                    {"label":"10-2008","values":[
                                                  {"label":"Syria","values":528},
                                                  {"label":"Afghanistan","values":688}
                                                  ]},
                    {"label":"11-2008","values":[
                                                  {"label":"Afghanistan","values":604},
                                                  {"label":"Syria","values":372}
                                                  ]},
                    {"label":"12-2008","values":[
                                                  {"label":"Syria","values":384},
                                                  {"label":"Afghanistan","values":800}
                                                  ]},
                    {"label":"01-2009","values":[
                                                  {"label":"Afghanistan","values":1648},
                                                  {"label":"Syria","values":436}
                                                  ]},
                    {"label":"02-2009","values":[
                                                  {"label":"Syria","values":532},
                                                  {"label":"Afghanistan","values":1204}
                                                  ]},
                    {"label":"03-2009","values":[
                                                  {"label":"Afghanistan","values":1024},
                                                  {"label":"Syria","values":372}
                                                  ]},
                    {"label":"04-2009","values":[
                                                  {"label":"Syria","values":384},
                                                  {"label":"Afghanistan","values":872}
                                                  ]},
                    {"label":"05-2009","values":[
                                                  {"label":"Afghanistan","values":1052},
                                                  {"label":"Syria","values":404}
                                                  ]},
                    {"label":"06-2009","values":[
                                                  {"label":"Syria","values":368},
                                                  {"label":"Afghanistan","values":1276}
                                                  ]},
                    {"label":"07-2009","values":[
                                                  {"label":"Afghanistan","values":1804},
                                                  {"label":"Syria","values":504}
                                                  ]},
                    {"label":"08-2009","values":[
                                                  {"label":"Syria","values":436},
                                                  {"label":"Afghanistan","values":2024}
                                                  ]},
                    {"label":"09-2009","values":[
                                                  {"label":"Afghanistan","values":2188},
                                                  {"label":"Syria","values":484}
                                                  ]},
                    {"label":"10-2009","values":[
                                                  {"label":"Syria","values":708},
                                                  {"label":"Afghanistan","values":2284}
                                                  ]},
                    {"label":"11-2009","values":[
                                                  {"label":"Afghanistan","values":2580},
                                                  {"label":"Syria","values":568}
                                                  ]},
                    {"label":"12-2009","values":[
                                                  {"label":"Syria","values":428},
                                                  {"label":"Afghanistan","values":2320}
                                                  ]},
                    {"label":"01-2010","values":[
                                                  {"label":"Afghanistan","values":2768},
                                                  {"label":"Syria","values":512}
                                                  ]},
                    {"label":"02-2010","values":[
                                                  {"label":"Syria","values":456},
                                                  {"label":"Afghanistan","values":2544}
                                                  ]},
                    {"label":"03-2010","values":[
                                                  {"label":"Afghanistan","values":2416},
                                                  {"label":"Syria","values":520}
                                                  ]},
                    {"label":"04-2010","values":[
                                                  {"label":"Syria","values":452},
                                                  {"label":"Afghanistan","values":2056}
                                                  ]},
                    {"label":"05-2010","values":[
                                                  {"label":"Afghanistan","values":1844},
                                                  {"label":"Syria","values":540}
                                                  ]},
                    {"label":"06-2010","values":[
                                                  {"label":"Syria","values":624},
                                                  {"label":"Afghanistan","values":2248}
                                                  ]},
                    {"label":"07-2010","values":[
                                                  {"label":"Afghanistan","values":2960},
                                                  {"label":"Syria","values":612}
                                                  ]},
                    {"label":"08-2010","values":[
                                                  {"label":"Syria","values":832},
                                                  {"label":"Afghanistan","values":3416}
                                                  ]},
                    {"label":"09-2010","values":[
                                                  {"label":"Afghanistan","values":2284},
                                                  {"label":"Syria","values":684}
                                                  ]},
                    {"label":"10-2010","values":[
                                                  {"label":"Syria","values":648},
                                                  {"label":"Afghanistan","values":2268}
                                                  ]},
                    {"label":"11-2010","values":[
                                                  {"label":"Afghanistan","values":1988},
                                                  {"label":"Syria","values":520}
                                                  ]},
                    {"label":"12-2010","values":[
                                                  {"label":"Syria","values":604},
                                                  {"label":"Afghanistan","values":2384}
                                                  ]}]
    , pc = piechart()
                    .width(500)
                    .height(400)
                    .data(singleData);

    var caller = _.bind(pc.init, pc);
d3.select('.js-pie-chart')
    .call(caller);

     // bg = bargraph()
     //                .width(700)
     //                .height(500)
     //                // .valuesPosition('fit')
     //                .barLayout('stacked')
     //                .anchor('bottom')
     //                .margin(0,0,50,50)
     //                .data(groupedData);




    // var axis = bg.axis()

    // axis.scale(bg.getOrdinalScale())
    //     .align('bottom')
    //     .ticks('auto', null)
    //     .show(true);

                  // .align('bottom')
                  // .show(true);

  // axis.scale()
/*
var caller = _.bind(bg.init, bg);

d3.select('.js-bar-chart')
    .call(caller);

// margins
d3.select('.js-margin-button').on('click', function (d) {
  var top = d3.select('.js-margin-top').node().valueAsNumber
    , right = d3.select('.js-margin-right').node().valueAsNumber
    , bottom = d3.select('.js-margin-bottom').node().valueAsNumber
    , left = d3.select('.js-margin-left').node().valueAsNumber
  ;

  bg.margin(top,right,bottom,left).draw();
  console.log(d, arguments);
});

// x ticks
d3.select('.js-x-ticks-button').on('click', function (d) {
  var numTicks = d3.select('.js-x-ticks').node().value
  ;

  bg.axis().ticks(numTicks).draw();
  console.log(numTicks);
});


// direction
d3.select('.js-direction-select').on('change', function (d) {
  // var dir = d3.select('.js-direction-select').node().valueAsNumber
  var dir = this.children[this.selectedIndex].value
    , axisDir = dir === 'bottom' || dir === 'top' ? 'left' : 'bottom';


  bg.anchor(dir).axis().align({linear:axisDir, ordinal: dir}).draw();
  console.log(d, arguments);
});

// dimensions
d3.select('.js-dimension-button').on('click', function (d) {
  var width = d3.select('.js-width').node().valueAsNumber
    , height = d3.select('.js-height').node().valueAsNumber
  ;

  bg.height(height).width(width).draw();
  console.log(d, arguments);
});


// rotation
d3.select('.js-rotation-button').on('click', function (d) {
  var linear = d3.select('.js-linear-rotate').node().valueAsNumber
    , ordinal = d3.select('.js-ordinal-rotate').node().valueAsNumber
  ;

  bg.axis().rotate({linear: linear, ordinal: ordinal}).draw();
  console.log(d, arguments);
});

*/


});